### PR TITLE
Created rules for the ELF BEACON sample shared by Intezer

### PR DESCRIPTION
**Pull request by ramen0x3f**
Some rules leveraging the newly added ELF support in capa.

### CODE REVIEW

File: host-interaction/hardware/memory/manipulate-linux-process-memory.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+rule:
+  meta:
+    name: manipulate Linux process memory
```
Comment (williballenthin) [inline host-interaction/hardware/memory/manipulate-linux-process-memory.yml:3]: the APIs referenced here are not linux specific (they're part of the c standard library) so lets loosen that restriction

File: host-interaction/hardware/memory/manipulate-linux-process-memory.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    - and:
+      - os: linux
+      - or:
+        - function:
+          - or:
+            - api: malloc
+            - api: memcpy
+            - api: memmove
+            - api: memset
+        - import: free
```
Comment (williballenthin) [inline host-interaction/hardware/memory/manipulate-linux-process-memory.yml:19]: i think we should prefer API functions to imports, unless the feature makes more sense at the file level. when we use an API feature, capa can tell us exactly which function has the capability. when we use imports, we just get a general sense of the capabilities.

so, lets make this a function (or basic block) scoped rule and rely on API features

File: host-interaction/mutex/create-semaphore.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+rule:
+  meta:
+    name: create semaphore
```
Comment (williballenthin) [inline host-interaction/mutex/create-semaphore.yml:3]:
```suggestion
    name: create semaphore on Linux
```

File: host-interaction/mutex/create-semaphore.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    namespace: host-interaction/mutex
+    author: alyssa.rahman@fireeye.com
+    scope: file
+    examples:
+      - 3db3e55b16a7b1b1afb970d5e77c5d98
+  features:
+    - and:
+      - os: linux
+      - import: sem_init
```
Comment (williballenthin) [inline host-interaction/mutex/create-semaphore.yml:12]:
```suggestion
      - api: sem_init
```

and change scope to bb or function

File: host-interaction/mutex/create-semaphore.yml (hunk 3 above)
Comment (williballenthin) [reply host-interaction/mutex/create-semaphore.yml:3]: same comment on other semaphore rules

File: host-interaction/mutex/create-semaphore.yml (hunk 4 above)
Comment (williballenthin) [reply host-interaction/mutex/create-semaphore.yml:12]: same comment on other semaphore rules

File: host-interaction/network/create-network-socket.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      - 3db3e55b16a7b1b1afb970d5e77c5d98
+  features:
+    - and:
+      - os: linux
+      - or:
+        - import: recv
+        - import: recvfrom
+        - import: setsockopt
+        - import: socket
+        - import: sendto
```
Comment (williballenthin) [inline host-interaction/network/create-network-socket.yml:17]: i think these routines are referenced in more specific ways by the rules found here: https://github.com/fireeye/capa-rules/tree/master/communication/socket

File: host-interaction/process/create/execute-command-through-syscall.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    name: execute command through syscall
+    namespace: host-interaction/process/create
+    author: alyssa.rahman@fireeye.com
+    scope: file
+    examples:
+      - 3db3e55b16a7b1b1afb970d5e77c5d98
+  features:
+    - and:
+      - os: linux
+      - import: syscall
```
Comment (williballenthin) [inline host-interaction/process/create/execute-command-through-syscall.yml:12]: `syscall` is an API that directly invokes the system call provided as an argument (i think, as documented here: https://man7.org/linux/man-pages/man2/syscall.2.html). so, this rule needs more specificity to disambiguate execution of a command from the other system calls (like recv, write, etc.).

File: lib/delay-execution.yml [hunk 7]
```diff
@@ -14,17 +16,23 @@ rule:
...
+              - api: kernel32.WaitForMultipleObjects
+              - api: kernel32.WaitForMultipleObjectsEx
+              - api: kernel32.RegisterWaitForSingleObject
+              - api: WaitOnAddress
+              - api: user32.MsgWaitForMultipleObjects
+              - api: user32.MsgWaitForMultipleObjectsEx
+              - api: NtDelayExecution
+              - api: KeWaitForSingleObject
+              - api: KeDelayExecutionThread
+      - import: signal
```
Comment (williballenthin) [inline lib/delay-execution.yml:37]: `signal` is used to configure the handling of many types of notifications, beyond sleeping (i think, at least as documented here: https://man7.org/linux/man-pages/man2/signal.2.html). so, i think this portion needs more logic to specify the specific type of signal that is being configured to effect a delay.

File: lib/delay-execution.yml [hunk 8]
```diff
@@ -14,17 +16,23 @@ rule:
...
+              - api: kernel32.WaitForSingleObjectEx
+              - api: kernel32.WaitForMultipleObjects
+              - api: kernel32.WaitForMultipleObjectsEx
+              - api: kernel32.RegisterWaitForSingleObject
+              - api: WaitOnAddress
+              - api: user32.MsgWaitForMultipleObjects
+              - api: user32.MsgWaitForMultipleObjectsEx
+              - api: NtDelayExecution
+              - api: KeWaitForSingleObject
+              - api: KeDelayExecutionThread
```
Comment (williballenthin) [inline lib/delay-execution.yml:36]: these should be restricted with `os: windows` while the other block restricted with `os: linux`. we haven't quite gotten to all the rules to specify the os, so when we come across them during updates, we should add those conditions in to make clear which APIs apply to which OSes.

**Review by williballenthin — CHANGES_REQUESTED** [review]
thank you @ramen0x3f! i'm excited that you're contributing these eight new rules that identify behaviors found in a real (interesting) malware sample.

as you can see from my inline comments, my biggest request is to prefer to use the "most specific scope" possible when identifying features, as this gives us the most information about where and how a behavior is referenced. with this in mind, please use `api` features at the BB or function scope rather than `import` features at the file scope, when possible. this will, e.g., let us associated capa matches with symfuncs or other function level systems.

feel free to ask any questions or for clarification on any of these points - happy to help get this stuff merged in!

File: host-interaction/hardware/memory/manipulate-linux-process-memory.yml (hunk 1 above)
Comment (mr-tz) [reply host-interaction/hardware/memory/manipulate-linux-process-memory.yml:3]: I wonder how useful this rule will be as one of these apis will be used in 99% of samples? I vote to remove this generic rule.

File: host-interaction/mutex/create-semaphore.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+rule:
+  meta:
+    name: create semaphore
+    namespace: host-interaction/mutex
+    author: alyssa.rahman@fireeye.com
+    scope: file
```
Comment (mr-tz) [inline host-interaction/mutex/create-semaphore.yml:6]:
```suggestion
    scope: function
```

Comment (mr-tz) [reply host-interaction/mutex/create-semaphore.yml:6]: see Willi's comment below

File: host-interaction/mutex/lock-semaphore.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    author: alyssa.rahman@fireeye.com
+    scope: file
+    examples:
+      - 3db3e55b16a7b1b1afb970d5e77c5d98
+  features:
+    - and:
+      - os: linux
+      - or:
+        - import: sem_timedwait
+        - import: sem_wait
```
Comment (mr-tz) [inline host-interaction/mutex/lock-semaphore.yml:14]: the linter will complain about a missing empty line at the end of the file, ensure to run `lint.py`/`lint.py --thorough` on all rule files

File: host-interaction/network/create-network-socket.yml (hunk 5 above)
Comment (mr-tz) [reply host-interaction/network/create-network-socket.yml:17]: agreed, vote to remove this generic rule

File: lib/delay-execution.yml [hunk 11]
```diff
@@ -14,17 +16,23 @@ rule:
...
-      - api: WaitOnAddress
-      - api: user32.MsgWaitForMultipleObjects
-      - api: user32.MsgWaitForMultipleObjectsEx
-      - api: NtDelayExecution
-      - api: KeWaitForSingleObject
-      - api: KeDelayExecutionThread
+      - function:
+        - and:
+          - basic block:
+            - or:
```
Comment (mr-tz) [inline lib/delay-execution.yml:22]: our current preference is to limit the scope via the file meta data, so I'd keep this as is.

**Comment by mr-tz** [comment]
Thanks, Alyssa, I've taken your work, made changes, and opened a new PR at #654.

**Pull request closed without merging**